pkg/sql/catalog/randgen/randgen.go

```diff
@@ -0,0 +1,331 @@
+			return pgerror.Newf(pgcode.InsufficientPrivilege,
+				"only admin users can generate more than %d descriptors at a time", maxNewDescsPerNonAdmin)
+		}
+
+		// Where are we creating the objects?
+		g.lookupBaseNamespaces(ctx, tn)
+
+		// Which templates are we using?
+		g.loadTemplates(ctx)
+
+		g.reset()
+		g.maybeGenDatabases(ctx)
+		g.flush(ctx)
+
+		if expDbs != g.cfg.GeneratedCounts.Databases ||
+			expScs != g.cfg.GeneratedCounts.Schemas ||
+			expTbs != g.cfg.GeneratedCounts.Tables {
+			return errors.AssertionFailedf("generated count mistmatch: got (%d,%d,%d), expected (%d,%d,%d)",
+				g.cfg.GeneratedCounts.Databases,
+				g.cfg.GeneratedCounts.Schemas,
+				g.cfg.GeneratedCounts.Tables,
+				expDbs, expScs, expTbs,
+			)
+		}
+
+		return nil
+	}()
+	return g.cfg, err
+}
+
+type genError struct {
+	err error
+}
+
+// Catalog interfaces with the schema resolver and privilege checker.
+type Catalog interface {
+	HasAdminRole(context.Context) (bool, error)
+	CheckAnyPrivilege(context.Context, privilege.Object) error
+	CanCreateDatabase(context.Context) error
+	CheckPrivilegeForUser(context.Context, privilege.Object, privilege.Kind, username.SQLUsername) error
+	ExpandTableGlob(context.Context, tree.TablePattern) (tree.TableNames, descpb.IDs, error)
+}
+
+// testSchemaGenerator encapsulates the state of the generator.
+type testSchemaGenerator struct {
+	cfg TestSchemaGeneratorConfig
+
+	txn *kv.Txn
+
+	// Catalog interface.
+	cat Catalog
+
+	// Session parameters.
+	user              username.SQLUsername
+	currentDatabase   string
+	currentSearchPath sessiondata.SearchPath
+	kvTrace           bool
+
+	// Descriptor collection and ID generator.
+	coll  *descs.Collection
+	idGen eval.DescIDGenerator
+
+	// The random number generator we're using.
+	rand *rand.Rand
+
+	// Which databases to generate or use.
+	createDatabases bool
+	numDatabases    int
+	dbNamePat       string
+	baseDatabase    *dbdesc.Mutable
+
+	// Which schemas to generate or use.
+	createSchemas            bool
+	numSchemasPerDatabase    int
+	scNamePat                string
+	useGeneratedPublicSchema bool
+	baseSchema               catalog.SchemaDescriptor
+
+	// Which tables to generate.
+	createTables       bool
+	numTablesPerSchema int
+	tbNamePat          string
+
+	// Template for the various descriptors being created. We define
+	// them here and then write to them in-place in each generation
+	// function so as to avoid heap allocations.
+	dbTemplate       descpb.DatabaseDescriptor
+	publicScTemplate descpb.SchemaDescriptor
+	scTemplate       descpb.SchemaDescriptor
+
+	templates []tbTemplate
+
+	// b is the current batch.
+	b *kv.Batch
+	// newDescs are the recently added descriptors that need a new
+	// namespace entry.
+	newDescs []catalog.MutableDescriptor
+	// queuedDescs are the recently updated descriptors that need to be
+	// written to KV.
+	queuedDescs []catalog.MutableDescriptor
+}
```

Comment:
I would find it useful to group the fields in this struct according to their lifecycle/mutability, either by rearranging the lines in which they appear, or by introducing sub-structs. Some fields are part of the "generator state" so to speak (like `baseSchema`, `rand`), some are part of the generator output (`b` etc), some are immutable configuration values (`cfg`, `kvtrace` etc) and some are dependencies (`cat`, `coll`, etc).